Enable new navbar for plugins (#572)
nautobot/nautobot: nautobot/docs/plugins/development.md

| `jobs` | The dotted path to the list of Job classes (default: `jobs.jobs`) |
| `max_version` | Maximum version of Nautobot with which the plugin is compatible |
| `menu_items` | The dotted path to the list of menu items provided by the plugin (default: `navigation.menu_items`) |

> | `menu_tabs` | The dotted path to the list of menu tab items provided by the plugin (default: `navigation.menu_tabs`) |

Reviewer: Rather than adding a new, separate item to the plugin declaration, can we make it so that `menu_items` can be defined in either the existing style (`PluginMenu*`) or the new style (`NavMenu*`)? I think that would be a bit more elegant. It feels like it should be possible to do something to automatically translate the old style into the new style in the core code - something like:

```python
if isinstance(menu_items[0], PluginMenuItem):
    # translate old-style plugin menu definitions into the new nav-menu items and buttons
    new_menu_items = []
    new_menu_item_weight = 100
    for menu_item in menu_items:        
        new_menu_button_weight = 100
        new_menu_buttons = []
        for button in menu_item.buttons:
            new_menu_buttons.append(NavMenuButton(
                link=button.link, title=button.title, icon_class=button.icon_class, permissions=button.permissions, weight=new_menu_button_weight
            ))
            new_menu_button_weight += 100

        new_menu_items.append(NavMenuItem(
            link=menu_item.link, link_text=menu_item.link_text, permissions=menu_item.permissions, weight=new_menu_item_weight, buttons=new_menu_buttons
        ))
        new_menu_item_weight += 100

    # wrap bare item/button list into the default "Plugins" menu tab and appropriate grouping
    menu_items = (
        NavMenuTab(name="Plugins", weight=..., groups=(
            NavMenuGroup(name=pluginconfig.verbose_name, weight=..., items=new_menu_items),
        ),
    )
```

Basically, let's think about the current "two approaches" situation as something temporary only - any groundwork we can lay now towards converging the two towards a single solution based around the NavMenu code is worth doing.

!!! note

>     Any buttons associated within a menu item will be shown only if the user has permission to view the link, regardless of what permissions are set on the buttons.

Reviewer: If this note is correct (I believe it is) it should probably be included in `navigation-menu.md`.

## Modifying Existing NavBar

Plugins can modify the existing navigation bar layout by defining `menu_items` inside of `navigation.py`. Using the key and weight system, a developer can integrate the plugin amongst existing menu tabs, groups, items and buttons and/or create entirely new menus as desired.

More documentation and examples can be found [here](../development/navigation-menu.md)

> 

Reviewer: This section should be integrated into (and perhaps replace entirely) the above `Navigation Menu Items` section.

!!! note
    Any buttons associated within a menu item will be shown only if the user has permission to view the link, regardless of what permissions are set on the buttons.

## Navigation Menu Tabs

> Plugins can modify the existing navigation bar layout by using `nav_tabs` inside of `navigation.py`. Using the key and weight system, a developer can integrate the plugin amongst existing menu tabs, groups, items and buttons and/or create entirely new menus as desired.

Reviewer: ```suggestion
Plugins can modify the existing navigation bar layout by defining `menu_items` inside of `navigation.py`. Using the key and weight system, a developer can integrate the plugin amongst existing menu tabs, groups, items and buttons and/or create entirely new menus as desired.
```

> ### Adding a new tab

Reviewer: Thinking about it, this whole section of documentation is now really applicable to both plugins and core apps, isn't it? I think it might be more appropriate to create a new document under `docs/development/` and link to it from the plugin docs as needed.

!!! note
    Any buttons associated within a menu item will be shown only if the user has permission to view the link, regardless of what permissions are set on the buttons.

## Navigation Menu Tabs

> Plugins can modify the existing navigation bar layout by using `nav_tabs` inside of `navigation.py`. Using the key and weight system, a developer can modify existing tabs, groups, items and buttons.

Reviewer: ```suggestion
Plugins can modify the existing navigation bar layout by using `nav_tabs` inside of `navigation.py`. Using the key and weight system, a developer can integrate the plugin amongst existing menu tabs, groups, items and buttons and/or create entirely new menus as desired..
```

### Adding a new tab

The code below shows how to add a new tab to the navbar. A tab is defined by a `NavMenuTab` object. Similarly a group is defined using `NavMenuGroup`. Both of these objects are used as containers for actual items.

> The position in the navigation menu is defined by the weight. The lower the weight the closer to the same the object will be. All core objects have weights in multiples of 100., meaning there is plenty of space around the objects for plugins to customize.

Reviewer: ```suggestion
The position in the navigation menu is defined by the weight. The lower the weight the closer to the start of the menus the object will be. All core objects have weights in multiples of 100, meaning there is plenty of space around the objects for plugins to customize.
```

> Below you can see `Dummy Tab` has a weight value of `150`. This means the tab will appear between `Organization` and `Devices`.

Reviewer: Do we have any straightforward way to introspect the current menu tab/group/item/button weights so as to figure out what value a plugin *should* use, short of digging through the code and reading each individual `navigation.py` file?

As an idea, maybe extend the nav-menu rendering template to add the weight of each tab/group/item/button as an HTML element [data attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) that could be inspected via the browser's dev-tools?

Author: Nice idea

Reviewer: Thanks for adding that. Mind also adding some notes here about how this information is available and can be used to figure out the desired weight?

Reviewer: Ah, I see you added that below already  -thanks! Might be better to move it up to here though?

Author: will move now

Author: actually does it need to be moved if we are now storing navbar documentation in `navigation-menu.md`?

Reviewer: See my new comments. :-)

                                    "dummy_plugin.add_dummymodel",
                                ],
                            ),
                            NavMenuImportButton(
                                link="plugins:dummy_plugin:dummymodel_import",
                                permissions=[
                                    "dummy_plugin.add_dummymodel"
                                ],
                            ),
                        ),
                    ),
                ),
            ),
        ),
    ),
)
```

### Modifying existing tab

By defining an object with the same identifier, a developer can modify existing objects. The example below shows modifying an existing tab to have a new group.

A tab object is being created with the same identifer as an existing object using the `name` attribute. Then a group is being created with a weight of `150`, which means it will appear between the already defined `Circuits` and `Provider` groups.

This pattern works for modifying all objects in the tree. New items can be added to groups and new buttons can be added to items.

``` python
menu_tabs = (
    NavMenuTab(
        name="Circuits",

>         groups=(

Reviewer: Thinking about how things were implemented in #485, I think this will actually result with errors being logged from `check_or_create_entry`, as presently implemented, because the `weight` of this `NavMenuTab` doesn't match the weight previously defined for `Circuits`.

 I do think we want to permit this approach (you shouldn't need to know the weight of an existing tab/group/item in order to augment it) but I think it will require some code changes to do so.

Author: This feature works. Tested in the dummy plugin

``` python
    NavMenuTab(
        name="Circuits",
        groups=(
            NavMenuGroup(
                name="Dummy Circuit Group",
                weight=150,
                items=(
                    NavMenuItem(
                        link="plugins:dummy_plugin:dummymodel_list",
                        link_text="Dummy Model",
                        permissions=["dummy_plugin.view_dummymodel"],
                        buttons=(
                            NavMenuAddButton(
                                link="plugins:dummy_plugin:dummymodel_add",
                                permissions=[
                                    "dummy_plugin.add_dummymodel",
                                ],
                            ),
                            NavMenuImportButton(
                                link="plugins:dummy_plugin:dummymodel_import",
                                permissions=["dummy_plugin.add_dummymodel"],
                            ),
                        ),
                    ),
                ),
            ),
        ),
    ),
```

Reviewer: And you confirmed that no `error` or `warning` messages are logged to the nautobot logs in this case? I'd really expect that we'd be hitting https://github.com/nautobot/nautobot/blob/develop-1.1.0/nautobot/core/apps/__init__.py#L43 in this scenario.

Author: No errors or warnings

Reviewer: Ok, can you help me understand why this is the case? Here's my understanding of why an error *should* be logged:

-  `NavMenuTab(name="Circuits", groups=...)` will default to a `weight` of `1000`.
- This is different than the prior declaration of the Circuits tab in `circuits/navigation.py` (which explicitly sets `weight=500`. 
- When we process the plugin `menu_items`, we call `create_or_check_entry()` on the `NavMenuTab`, which includes `("weight", 1000)` amongst its `fixed_fields`.
- This doesn't match the previously registered weight of 500, so a `logger.error()` should be triggered.

Author: Ah i see whats happening. For some reason my log messages are not coming through my terminal anymore. There is an error happening, so when the plugin defines `circuits` the weight value defaults to `1000`. This logs an error saying the weight was originally set to `150` but for some reason this wasnt being showed in my `stdout`
